File: sd-calendar-events/src/calendar-helper.js
#!/usr/bin/env osascript -l JavaScript

// JXA script to read events from macOS Calendar.app
// Uses Automation permission (SD -> Calendar.app)

function run(argv) {
  var command = argv[0] || "events";
  var filterNames = argv[1] ? argv[1].split("||") : null;

  var app = Application("Calendar");

  var names = app.calendars.name();
  var count = names.length;

  // Get colors one by one (bulk access returns nested arrays inconsistently)
  var calendarInfos = [];
  for (var i = 0; i < count; i++) {
    var hex = "#888888";
    try {
      var c = app.calendars[i].color();
      if (c && c.length >= 3) {
        hex = rgbToHex(c);
      }
    } catch (e) {}
    calendarInfos.push({
      id: names[i],
      title: names[i],
      color: hex,
      source: ""
    });
  }

  if (command === "calendars") {
    return JSON.stringify({ events: [], calendars: calendarInfos, error: null });
  }

  // events command
  var now = new Date();
  // Use day boundaries just outside today so _greaterThan/_lessThan capture the full day
  var yesterday = new Date(now.getFullYear(), now.getMonth(), now.getDate() - 1, 23, 59, 59);
  var tomorrow = new Date(now.getFullYear(), now.getMonth(), now.getDate() + 1, 0, 0, 0);

  var events = [];

  for (var i = 0; i < count; i++) {
    var calName = names[i];
    if (filterNames && filterNames.indexOf(calName) < 0) continue;

    var cal = app.calendars[i];
    var calEvts;
    try {
      calEvts = cal.events.whose({
        _and: [
          { startDate: { _greaterThan: yesterday } },
          { startDate: { _lessThan: tomorrow } }
        ]
      })();
    } catch (e) {
      // Fallback: some calendars don't support whose, skip them
      continue;
    }

    for (var j = 0; j < calEvts.length; j++) {
      try {
        var ev = calEvts[j];
        var startDate = ev.startDate();
        var endDate = ev.endDate();
        var isAllDay = ev.alldayEvent();

        // Show all today's events (past ones show as "Ended")

        var summary = ev.summary() || "(No title)";
        var desc = safeGet(ev, "description");
        var loc = safeGet(ev, "location");
        var url = safeGet(ev, "url");

        // Build a stable-ish ID from summary + start time
        var id = calName + "|" + summary + "|" + startDate.getTime();

        events.push({
          id: id,
          title: summary,
          startDate: startDate.getTime(),
          endDate: endDate.getTime(),
          isAllDay: isAllDay,
          meetingUrl: extractMeetingUrl(url, loc, desc),
          calendarName: calName,
          calendarColor: calendarInfos[i] ? calendarInfos[i].color : "#888888"
        });
      } catch (e) {
        continue;
      }
    }
  }

  events.sort(function (a, b) { return a.startDate - b.startDate; });
  return JSON.stringify({ events: events, calendars: calendarInfos, error: null });
}

function safeGet(ev, prop) {
  try {
    var val = ev[prop]();
    return (val === null || val === undefined) ? "" : String(val);
  } catch (e) {
    return "";
  }
}

function extractMeetingUrl(url, location, description) {
  var pattern = /https?:\/\/[^\s<"')]*(?:zoom\.us|teams\.microsoft\.com|meet\.google\.com|webex\.com)[^\s<"')"]*/i;
  if (url && pattern.test(url)) return url.match(pattern)[0];
  if (location) { var m = location.match(pattern); if (m) return m[0]; }
  if (description) { var m = description.match(pattern); if (m) return m[0]; }
  if (url && /^https?:\/\//.test(url)) return url;
  return null;
}

function rgbToHex(c) {
  if (!c || !Array.isArray(c) || c.length < 3) return "#888888";
  // Calendar.app returns 0.0-1.0 float RGB
  var rr = Math.round(c[0] * 255);
  var gg = Math.round(c[1] * 255);
  var bb = Math.round(c[2] * 255);
  return "#" + ((1 << 24) + (rr << 16) + (gg << 8) + bb).toString(16).slice(1).toUpperCase();
}

File: sd-calendar-events/src/swift/CalendarHelper.swift
import EventKit
import Foundation

struct EventOutput: Codable {
    let id: String
    let title: String
    let startDate: Double
    let endDate: Double
    let isAllDay: Bool
    let meetingUrl: String?
    let calendarName: String
    let calendarColor: String
}

struct Output: Codable {
    let events: [EventOutput]
    let calendars: [CalendarInfo]
    let error: String?
}

struct CalendarInfo: Codable {
    let id: String
    let title: String
    let color: String
    let source: String
}

let store = EKEventStore()

func hexColor(_ cgColor: CGColor?) -> String {
    guard let c = cgColor, let rgb = c.converted(to: CGColorSpaceCreateDeviceRGB(), intent: .defaultIntent, options: nil),
          let comps = rgb.components, comps.count >= 3 else { return "#888888" }
    let r = Int(comps[0] * 255)
    let g = Int(comps[1] * 255)
    let b = Int(comps[2] * 255)
    return String(format: "#%02X%02X%02X", r, g, b)
}

func extractMeetingUrl(from event: EKEvent) -> String? {
    if let url = event.url?.absoluteString,
       url.contains("zoom.us") || url.contains("teams.microsoft.com") || url.contains("meet.google.com") || url.contains("webex.com") {
        return url
    }
    if let loc = event.location {
        if let match = loc.range(of: #"https?://[^\s<"')]*(?:zoom\.us|teams\.microsoft\.com|meet\.google\.com|webex\.com)[^\s<"')]*"#, options: .regularExpression) {
            return String(loc[match])
        }
    }
    if let notes = event.notes {
        if let match = notes.range(of: #"https?://[^\s<"')]*(?:zoom\.us|teams\.microsoft\.com|meet\.google\.com|webex\.com)[^\s<"')]*"#, options: .regularExpression) {
            return String(notes[match])
        }
    }
    if let url = event.url?.absoluteString {
        return url
    }
    return nil
}

/// Request access interactively — only works from Terminal where macOS can show the prompt.
func requestAccessInteractively() -> Bool {
    let semaphore = DispatchSemaphore(value: 0)
    var granted = false

    if #available(macOS 14.0, *) {
        store.requestFullAccessToEvents { g, _ in
            granted = g
            semaphore.signal()
        }
    } else {
        store.requestAccess(to: .event) { g, _ in
            granted = g
            semaphore.signal()
        }
    }

    // Timeout after 30s — prevents hanging when launched from a non-interactive context
    let result = semaphore.wait(timeout: .now() + 30)
    if result == .timedOut {
        return false
    }
    return granted
}

/// Check if we have calendar access without prompting.
func checkAccess() -> (authorized: Bool, needsPrompt: Bool) {
    let status = EKEventStore.authorizationStatus(for: .event)
    if #available(macOS 14.0, *) {
        switch status {
        case .fullAccess:
            return (true, false)
        case .notDetermined:
            return (false, true)
        default:
            return (false, false)
        }
    } else {
        switch status {
        case .authorized:
            return (true, false)
        case .notDetermined:
            return (false, true)
        default:
            return (false, false)
        }
    }
}

func run() {
    let command = CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : "events"

    // "authorize" command — meant to be run from Terminal to trigger the permission prompt
    if command == "authorize" {
        let (authorized, needsPrompt) = checkAccess()
        if authorized {
            print("{\"error\":null,\"events\":[],\"calendars\":[]}")
            fputs("✓ Calendar access already granted.\n", stderr)
            return
        }
        if needsPrompt {
            fputs("Requesting calendar access — look for the macOS permission prompt...\n", stderr)
            let granted = requestAccessInteractively()
            if granted {
                fputs("✓ Calendar access granted!\n", stderr)
                print("{\"error\":null,\"events\":[],\"calendars\":[]}")
            } else {
                fputs("✗ Calendar access denied or timed out.\n", stderr)
                fputs("  Go to: System Settings → Privacy & Security → Calendars\n", stderr)
                fputs("  Grant access to 'calendar-helper' or 'Elgato Stream Deck'\n", stderr)
                print("{\"error\":\"Calendar access denied. Grant in System Settings → Privacy & Security → Calendars.\",\"events\":[],\"calendars\":[]}")
            }
        } else {
            fputs("✗ Calendar access is denied.\n", stderr)
            fputs("  Go to: System Settings → Privacy & Security → Calendars\n", stderr)
            fputs("  Grant access to 'calendar-helper' or 'Elgato Stream Deck'\n", stderr)
            print("{\"error\":\"Calendar access denied. Grant in System Settings → Privacy & Security → Calendars.\",\"events\":[],\"calendars\":[]}")
        }
        return
    }

    // For non-interactive commands (events/calendars), check status without prompting
    let (authorized, needsPrompt) = checkAccess()
    if !authorized {
        if needsPrompt {
            // Try requesting — might work, might not, but with a timeout so we don't hang
            let granted = requestAccessInteractively()
            if !granted {
                printJSON(Output(events: [], calendars: [], error: "NOT_AUTHORIZED: Run from Terminal first: ./calendar-helper authorize"))
                return
            }
        } else {
            printJSON(Output(events: [], calendars: [], error: "ACCESS_DENIED: Grant access in System Settings → Privacy & Security → Calendars"))
            return
        }
    }

    let filterIds: Set<String>? = CommandLine.arguments.count > 2
        ? Set(CommandLine.arguments[2].split(separator: ",").map(String.init))
        : nil

    let allCalendars = store.calendars(for: .event)

    let calendarInfos = allCalendars.map { cal in
        CalendarInfo(id: cal.calendarIdentifier, title: cal.title, color: hexColor(cal.cgColor), source: cal.source.title)
    }

    if command == "calendars" {
        printJSON(Output(events: [], calendars: calendarInfos, error: nil))
        return
    }

    // Fetch today's events
    let now = Date()
    let calendar = Calendar.current
    let endOfDay = calendar.date(bySettingHour: 23, minute: 59, second: 59, of: now)!

    let calendarsToSearch = filterIds != nil
        ? allCalendars.filter { filterIds!.contains($0.calendarIdentifier) }
        : allCalendars

    let predicate = store.predicateForEvents(withStart: now, end: endOfDay, calendars: calendarsToSearch)
    let ekEvents = store.events(matching: predicate).sorted { $0.startDate < $1.startDate }

    let events = ekEvents.map { ev in
        EventOutput(
            id: ev.eventIdentifier ?? UUID().uuidString,
            title: ev.title ?? "(No title)",
            startDate: ev.startDate.timeIntervalSince1970 * 1000,
            endDate: ev.endDate.timeIntervalSince1970 * 1000,
            isAllDay: ev.isAllDay,
            meetingUrl: extractMeetingUrl(from: ev),
            calendarName: ev.calendar.title,
            calendarColor: hexColor(ev.calendar.cgColor)
        )
    }

    printJSON(Output(events: events, calendars: calendarInfos, error: nil))
}

func printJSON(_ output: Output) {
    let encoder = JSONEncoder()
    if let data = try? encoder.encode(output), let json = String(data: data, encoding: .utf8) {
        print(json)
    }
}

run()
